### ISSUE TITLE
Add EventLog interface

### ISSUE THREAD

**Issue by codeliner** [prooph/event-store#319]
2 days ago I presented a talk about our event store at PHP UserGroup Dresden.

While preparing the talk I thought about the key features a database system has to provide so that it can be used as a prooph event store. The four features I came up with:

- Transaction support: storing multiple events within a transaction
- JSON query support: Query event metadata which is stored as JSON
- Sequences (or Autoincrement): event stream positions
- Unique multikey indexes: [optimistic locking](http://slides.com/prooph/deck-0-1-4#/11)

I talked with @oqq about optimistic locking and if it is really needed. @oqq pointed out that in many cases you don't need optimistic locking but instead you simple merge concurrent operations on the same aggregate. If they happen in parallel then your application needs to deal with it instead of throwing a concurrency exception.

Last year I had a short chat with @Ocramius about Kafka and that we cannot provide a Kafka event store because optimistic locking on aggregate id + aggregate version isn't possible with Kafka. Marco had the same thoughts like Eric that you don't necessarily need locking.

So I'd like to bring up that topic for discussion here. A Kafka event store would help to increase visibility of prooph even more and the performance and concepts possible with Kafka would allow some interesting use cases for high load handled by distributed systems.

I'd also reconsider a mongoDB v7 event store in that case, because I think the other three features can be achieved with mongo and since mongoDB 3.6 there is this nice feature called [Change Streams](https://docs.mongodb.com/manual/reference/method/db.collection.watch/) which would be awesome for projections.

I'm currently thinking about 3 possible results of this discussion:

1. Keep everything as-is because we have Postgres and also ArangoDB in the near future which both offer all 4 features
2. Introduce an additional `EventLog` (or another name) interface which marks stores without optimistic locking capabilities
3. Just implement Kafka and MongoDB and only note in the README that optimistic locking is not possible when using these stores

**Comment by fritz-gerneth** [comment]
Possibly lacking the technical knowledge about this, for Kafka specifically. My assumption so far is that concurrent changes on the same can only be merged if they are not depending on each other. For two stateless DTOs I can think of how this does work.
But how would any generic implementation know if two events are unrelated two each other? By simply comparing the payload? This seems rather dangeous. The main problem I see is to merge the event while it has been submitted based on a old state, while the new event might not have caused the event. 
So the question is, how to merge concurrent events without knowing their dependencies and business logic?  Again there is a good chance there is a solution for this, I'm not the expert on this and would really love to read on this approach if it exists :) 

Putting this aside I think the ability to merge concurrent events is an implementation detail of the event store implementation. If the ES implementation can handle (some) conflicts, it could do so. For everything else we still have the exception delegating the resolution to the application (as it should). In the end my application only cares about concurrency exceptions that could not be handled. If some get handled automatically, all the better. 
If someone does not like ths behavior, there is always the option to make this configurable (for an implementation).

**Comment by codeliner** [comment]
@fritz-gerneth When using Kafka you have to ensure consistency on the producer side (in the aggregate for example). Having said this I'm not suggesting that our event store should try to merge events but that this behavior is delegated back to the domain model. 
In other words:
You have to use Postgres, MySql, MariaDB or ArangoDB if your aggregates require optimistic locking.

You can use Kafka or MongoDB if you don't need optimistic locking on database level and handle concurrency on application/domain level. One option would be pessimistic locking within the aggregate or simply except alle events as they occur and handle conflicts with a Saga or other types of supervisor process

**Comment by Ocramius** [comment]
@codeliner optimistic locking may not be needed at all. There are two alternatives:

 1. pessimistic locking - locking is performed upfront, by identifier, on a centralized or at least highly consistent system. No operations are allowed on an aggregate until the lock is available again.
 2. no locking at all - requires pure aggregates. Each aggregate only records events, and is not allowed to make operations fail (exceptions). Failures are also events. Only new events can happen, and event versioning is also gone from the aggregate, replaced by a vector clock or a centralized sequence to prevent collisions in event order. This can only be applied to aggregates where the domain can "heal" from concurrency issues.

**Comment by fritz-gerneth** [comment]
I see, did misread your comment on merging then, hoping for a solution for the general mergic problem :) 

As you said, it is actually up to the domain which locking mechanism it requires. So the locking strategy is determined not by the store but the domain. The store merely can say that it doesn't support this. Ideally this is even configurable on a per AR level and not globally for an ES?

**Comment by prolic** [comment]
Currently the concurrency exception is only thrown, based on the used persistence strategy. So this behaviour can be simply changed by using a custom strategy.

Pessimistic locking is a valid approach, although it has it's downsides.

No locking at all is only a valid approach, if that fits in the domain, in some cases this approach can be very harmful. So that's up to the user to either change the persistence strategy (or use the default provided ones).

About resolving concurrency conflicts, this is an issue that has to be solved within the domain, and sometimes that's not possible an a concurrency exception has to be thrown again. There is a nice blog-post which also includes some sample code: http://danielwhittaker.me/2014/09/29/handling-concurrency-issues-cqrs-event-sourced-system/

About the kafka event-store: I will give some more insights into this (probably next week), so I like to skip this topic for now.

About mongodb: Biggest problem is missing transaction support, there is no way around this, so it's only useful with some constraints (very rare concurrency issues, probably applications without high load). Auto-increment can be achieved manually somehow, but it's still an issue, because you can't do it within a transaction.

**Comment by codeliner** [comment]
@prolic For mongoDB I'm thinking of using documents as "chunks/partitions/segments". New events are added to the "active" document within a collection (the stream). These chunk documents have a sequence as `_id` which is incremented by one as soon as the active chunk is "full" (definition of full should be configurable).
If the active chunk is full a new chunk will be added with the incremented sequence number.
There is only one active chunk at the same time.
Every chunk also has a field `previousChunkEnd` which is the total stream position at the time the chunk was created. Event stream position of a single event within a chunk is calculated this way:
`chunk.previousChunkEnd + chunk.events.length` (something like that, not sure atm if mongo keeps the order of an array but I think so)

This way you get "transaction" support because single document operations are atomic.
Metadata queries should also be possible.
Only optimistic locking is a problem (in theory at least)

**Comment by codeliner** [comment]
To summarize my RFC:

I'd like to bring back mongoDB as a prooph event store. We're using it in of our production systems but cannot upgrade to event store v7 and hence cannot make use of all the other nice stuff like projections, event-store-http-api ...

I'd also like to add Kafka (looking forward to @prolic's blog post about the topic).

Because we have the optimistic locking for Postgres etc. my request is if we want to **a)** Introduce another interface to be able to distinguish between stores that can provide optimistic locking and the ones that don't or **b)** just put bold note in the docs so that people are aware of the limitation or **c)** keep those databases out of scope for prooph

**Comment by prolic** [comment]
As for now, I think option b) `just put a bold note in the docs` is good enough. About MongoDB as event-store, maybe the following could work:

event_stream:
```
_id: <some aggregate uuid>
data: {
    'event_1': <some mongo id>
    'event_2':  <some other mongo id>
    ...
},
version: <current version>
```

data_stream:
```
_id: <some mongo id>
payload: <some payload>
metadata: <some metadata>
created_at: <created at>
```

This way we would store only minimal information into the aggregate's document and probably never hit the 16MB limit, we could add some information to the docs that (has to be calculated, don't take this example!!!) only 100.000 events can be stored for a single aggregate. After that you either model smaller aggregates or create a new aggregate (once limit reached, every month, every week, ...) based on previous aggregate with a link the the former one.

Unique constraints can be added in MongoDB, so I don't see the optimistic locking as a problem right now (I may be wrong).

**Comment by prolic** [comment]
You can read also here about it: [Why there will be no Kafka EventStore in prooph](https://www.sasaprolic.com/2018/03/why-there-will-be-no-kafka-eventstore.html)
